nemo/collections/asr/models/ctc_models.py

```diff
@@ -662,14 +662,19 @@ def _transcribe_output_processing(self, outputs, trcfg: TranscribeConfig) -> Gen
         current_hypotheses, all_hyp = self.decoding.ctc_decoder_predictions_tensor(
             logits, decoder_lengths=logits_len, return_hypotheses=trcfg.return_hypotheses,
         )
-        logits = logits.cpu()
-
         if trcfg.return_hypotheses:
+            # See comment in
+            # ctc_greedy_decoding.py::GreedyCTCInfer::forward() to
+            # understand this idiom.
+            logits_cpu = torch.empty(logits.shape, dtype=logits.dtype, pin_memory=True)
```

Comment:
This is a very neat trick. Just a question - what happens if i execute this on an environment with no GPU at all (not even cuda is setup). Will it work fine?

Reply:
Reason being - CPU only inference is quite common with our ASR models, especially on Gradio / HF space.

Reply:
Finally, no need to set device=something? The copy_ automatically does device to host copy?

Reply:
> what happens if i execute this on an environment with no GPU at all (not even cuda is setup). Will it work fine?

You're right, it doesn't just work in a CPU-only environment. I get:

```
Python 3.9.18 (main, Sep 11 2023, 08:25:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.5.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: a = torch.empty((2,2,2), pin_memory=True)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In [2], line 1
----> 1 a = torch.empty((2,2,2), pin_memory=True)

RuntimeError: Pinned memory requires CUDA. PyTorch splits its backend into two shared libraries: a CPU library and a CUDA library; this error has occurred because you are trying to use some CUDA functionality, but the CUDA library has not been loaded by the dynamic linker for some reason.  The CUDA library MUST be loaded, EVEN IF you don't directly use any symbols from the CUDA library! One common culprit is a lack of -Wl,--no-as-needed in your link arguments; many dynamic linkers will delete dynamic library dependencies if you don't depend on any of their symbols.  You can check if this has occurred by using ldd on your binary to see if there is a dependency on *_cuda.so library.
```

Getting this to work in a sensible way without too many conditional statements (after all, in the existing code, .cpu() is a no-op in the case that the user was running on CPU; meanwhile, here, I would impose an extra memory copy on those users) is probably the main challenge with this PR.

> Finally, no need to set device=something?

I also made a mistake here, it turns out. I assumed that device=None would default to the "cpu", but it actually defaults to whatever you set with https://pytorch.org/docs/stable/generated/torch.set_default_device.html#torch.set_default_device So I need to explicitly use torch.device("cpu") here.